tangelo/linq/tests/test_simulator.py

```diff
@@ -636,6 +640,140 @@ def backend_info():
+        for backend in (installed_simulator | installed_clifford_simulators) - installed_cmeasure_simulators:
+            sim = get_backend(backend, n_shots=1)
+            self.assertRaises(RuntimeError, sim.simulate, cmeas_circuit)
+
+    def test_measurement_controlled_gates_function(self):
+        """Test that the repeat until success circuit is properly performed."""
+
+        for backend in installed_cmeasure_simulators:
+            sim = get_backend(backend, n_shots=1)
+
+            # Repeat until success H+CNOT to prepare |11>
+            def cfunc(measurement):
+                return [Gate("H", 0), Gate("CMEASURE", 0)] if measurement == "0" else []
+
+            applied_circuit = Circuit([Gate('H', 0),
+                                       Gate('CMEASURE', 0, parameter="0"), Gate('H', 0),
+                                       Gate('CMEASURE', 0, parameter="0"), Gate('H', 0),
+                                       Gate('CMEASURE', 0, parameter="1"),
+                                       Gate('CNOT', 1, control=0)])
+
+            # Test that the probability is correct when two failures followed by a success occur
+            circuit = Circuit([Gate("H", 0), Gate("CMEASURE", 0),  Gate("CNOT", 1, 0)], cmeasure_control=cfunc)
+            f, _ = sim.simulate(circuit, save_mid_circuit_meas=True, desired_meas_result="001")
+            assert_freq_dict_almost_equal(f, {"11": 1}, 1.e-7)
+            assert_freq_dict_almost_equal(circuit.success_probabilities, {"001": 0.125}, 1.e-7)
+            self.assertTrue(applied_circuit == Circuit(circuit.applied_gates))
+
+            # Test that the correct ordering of probabilities is obtained when running without specifying the
+            # desired measurement result.
+            sim.n_shots = 1000
+
+            circuit = Circuit([Gate("H", 0), Gate("CMEASURE", 0),  Gate("CNOT", 1, 0)], cmeasure_control=cfunc)
+            f, _ = sim.simulate(circuit, save_mid_circuit_meas=True)
+            assert_freq_dict_almost_equal(f, {"11": 1}, 1.e-7)
+            self.assertTrue(sim.mid_circuit_meas_freqs["1"] > sim.mid_circuit_meas_freqs["01"])
+            self.assertTrue(sim.mid_circuit_meas_freqs["01"] > sim.mid_circuit_meas_freqs["001"])
+
+    def test_measurement_controlled_gates_class(self):
+        """Test of an adaptive IterativeQPE circuit."""
+
+        for backend in installed_cmeasure_simulators:
+
+            sim = get_backend(backend, n_shots=2)
+
+            # Unitary has eigenvalue (1/16+1/8+1/2)*pi with eigenvector |1> on qubit 1.
+            # Measurement will be 0011010 = 0/64 + 0/32 + 1/16 + 1/8 + 0/4 + 1/2 + 0
+            ugate = Gate("CPHASE", 1, control=0, parameter=np.pi*(1/16 + 1/8 + 1/2))
+
+            class IterativeQPE(ClassicalControl):
+                def __init__(self, n_bits: int):
+                    """Iterative QPE with n_bits"""
+                    self.n_bits: int = n_bits
+                    self.bitplace: int = copy(self.n_bits)
```

Comment:
Why not just setting it to n_bits ? 

You do not need to copy anything here, n_bits (and self.n_bits) is immutable: it's an integer. Even if you changed the value of n_bits later in the body of this function, it wouldn't affect other variables.

```
class xy():
    
    def __init__(self):
        self.x = 1
        self.y = self.x
        
o = xy()

o.x = 2
print(o.x, o.y)   # prints 2 1
```